streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -842,7 +842,14 @@ public synchronized ProcessorTopology buildTopology() {
             nodeGroup.addAll(value);
         }
         nodeGroup.removeAll(globalNodeGroups());
-
+        for (final NodeFactory<?, ?> entry : nodeFactories.values()) {
+            if (entry instanceof ProcessorNodeFactory) {
+                ProcessorNodeFactory<?, ?> factory = (ProcessorNodeFactory<?, ?>) entry;
+                if (factory.supplier.get() == factory.supplier.get()) {
+                    throw new TopologyException("topology has singleton result of ProcessorSupplier " + factory.name);
+                }
+            }
+        }
```

Comment:
It might be best to extract this into a helper method and call directly when `addProcessor()` is called. That way the corresponding stack trace is easier to map to the actually call to `addProcessor()` that passed in incorrect supplier?

Similarly, we should add this check to `KStreamImpl#process()` and others

Reply:
That's a great idea. I introduced a util class `TopologyUtil` under processor package.